src/host/posix/multicast_routing_manager.hpp

```diff
@@ -0,0 +1,138 @@
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef BACKBONE_ROUTER_MULTICAST_ROUTING_MANAGER_HPP_
+#define BACKBONE_ROUTER_MULTICAST_ROUTING_MANAGER_HPP_
+
+#include "openthread-br/config.h"
+
+#include <set>
+
+#include <openthread/backbone_router_ftd.h>
+
+#include "common/code_utils.hpp"
+#include "common/mainloop.hpp"
+#include "common/time.hpp"
+#include "common/types.hpp"
+#include "host/posix/infra_if.hpp"
+#include "host/posix/netif.hpp"
+#include "host/thread_host.hpp"
+
+namespace otbr {
+
+class MulticastRoutingManager : public MainloopProcessor, private NonCopyable
+{
+public:
+    explicit MulticastRoutingManager(const Netif                   &aNetif,
+                                     const InfraIf                 &aInfraIf,
+                                     const Host::NetworkProperties &aNetworkProperties);
+
+    void Deinit(void) { FinalizeMulticastRouterSock(); }
+    bool IsEnabled(void) const { return mMulticastRouterSock >= 0; }
+    void HandleStateChange(otBackboneRouterState aState);
+    void HandleBackboneMulticastListenerEvent(otBackboneRouterMulticastListenerEvent aEvent,
+                                              const Ip6Address                      &aAddress);
+
+private:
+    enum
+    {
+        kUsPerSecond                              = 1000000, //< Microseoncs per second.
+        kMulticastForwardingCacheExpireTimeout    = 300, //< Expire timeout of Multicast Forwarding Cache (in seconds)
+        kMulticastForwardingCacheExpiringInterval = 60,  //< Expire interval of Multicast Forwarding Cache (in seconds)
+        kMulticastMaxListeners                    = 75,  //< The max number of Multicast listeners
+        kMulticastForwardingCacheTableSize        = kMulticastMaxListeners * 10, //< The max size of MFC table.
+    };
```

Comment:
I suggest using `static constexpr` constants instead of `enum`. The advantage is that the constant's type is well-defined.

We used `enum` early on before OpenThread moved to `C++11` (where `constexpr` is first introduced).